backend/src/v2/driver/driver.go

```diff
@@ -1850,10 +1897,46 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
 				subTaskName = v.ProducerSubtask
 				outputArtifactKey = v.OutputArtifactKey
 			}
-			// If the sub-task is a DAG, reassign currentTask and run
-			// through the loop again.
-			currentTask = tasks[subTaskName]
-			// }
+
+			if currentTask.GetExecution().GetCustomProperties()["iteration_count"] != nil {
+				glog.V(4).Infof("currentTask, %v, is a ParallelFor DAG. Attempting to gather all producer_subtask iterations of %w", currentTask.TaskName(), subTaskName)
+
+				outputs_list := make([]*pipelinespec.RuntimeArtifact, 0)
+				for i := range currentTask.GetExecution().GetCustomProperties()["iteration_count"].GetIntValue() {
+					// Follow the convention set in mlmd.GetExecutionsInDAG for ParallelFor tasks
+					subTaskIterationName := fmt.Sprintf("%s_idx_%v", subTaskName, i)
+					glog.V(4).Infof("subTaskIterationName: %v", subTaskIterationName)
+					subTask, ok := tasks[subTaskIterationName]
+					if !ok {
+						return nil, cfg.err(fmt.Errorf("subTaskName, %v, not in tasks", subTaskName))
+					}
+					outputArtifacts, err := cfg.mlmd.GetOutputArtifactsByExecutionId(cfg.ctx, subTask.GetID())
+					if err != nil {
+						return nil, err
+					}
+
+					runtimeArtifact, err := outputArtifacts[outputArtifactKey].ToRuntimeArtifact()
+					if err != nil {
+						cfg.err(err)
+					}
+					outputs_list = append(outputs_list, runtimeArtifact)
+				}
+				return &pipelinespec.ArtifactList{
```

Comment:
Could you update the comment `// Continue looping until we reach a sub-task that is NOT a DAG.`? It seems to apply anymore since ParallelFor is a DAG.